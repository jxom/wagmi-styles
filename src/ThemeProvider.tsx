import * as React from 'react';

export const ThemeContext = React.createContext({ theme: { _id: '_' } });

export const ThemeProvider = ({ children, theme }: any) => (
  <ThemeContext.Provider value={{ theme }}>{children}</ThemeContext.Provider>
);
