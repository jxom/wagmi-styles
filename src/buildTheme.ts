import { createContext } from 'react';
import { generateStyled } from './private/generateStyled';
import { generateStyleFn } from './private/generateStyleFn';
import { generateStyleVariantsFn } from './private/generateStyleVariantsFn';
import { generateThemeProvider } from './private/generateThemeProvider';
import {
  generateUseStyle,
  generateUseStyleVariants,
} from './private/generateUseStyle';
import type { ThemeDefinition, ThemeVars, Tokens } from './types';

export function buildTheme<ThemeTokens extends Tokens>(
  ...themes: Array<ThemeDefinition<ThemeTokens>>
) {
  const ThemeContext = createContext({
    theme: { _id: themes?.[0]?._id || '_' },
  });

  const themeMap = (themes as Array<ThemeDefinition<ThemeTokens>>).reduce(
    (
      currentThemeMap: { [id: string]: ThemeVars<ThemeTokens> },
      theme: ThemeDefinition<ThemeTokens>
    ) => {
      return {
        ...currentThemeMap,
        [theme._id]: theme,
      };
    },
    {}
  );

  const style = generateStyleFn(
    themeMap as { [id: string]: ThemeVars<ThemeTokens> }
  );
  const styleVariants = generateStyleVariantsFn(
    themeMap as { [id: string]: ThemeVars<ThemeTokens> }
  );
  const ThemeProvider = generateThemeProvider(ThemeContext);
  const useStyle = generateUseStyle(ThemeContext);
  const useStyleVariants = generateUseStyleVariants(ThemeContext);
  const styled = generateStyled({ style, useStyle });

  return {
    style,
    styleVariants,
    styled,
    ThemeProvider,
    useStyle,
    useStyleVariants,
  };
}
