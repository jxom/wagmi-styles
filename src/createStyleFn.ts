import { generateStyleFn } from './private/generateStyleFn';
import { generateStyleVariantsFn } from './private/generateStyleVariantsFn';
import type { ThemeDefinition, ThemeVars, Tokens } from './types';

export function createStyleFn<ThemeTokens extends Tokens>(
  ...themes: Array<ThemeDefinition<ThemeTokens>>
) {
  const themeMap = (themes as Array<ThemeDefinition<ThemeTokens>>).reduce(
    (
      currentThemeMap: { [id: string]: ThemeVars<ThemeTokens> },
      theme: ThemeDefinition<ThemeTokens>
    ) => {
      return {
        ...currentThemeMap,
        [theme._id]: theme,
      };
    },
    {}
  );

  const style = generateStyleFn(
    themeMap as { [id: string]: ThemeVars<ThemeTokens> }
  );
  const styleVariants = generateStyleVariantsFn(
    themeMap as { [id: string]: ThemeVars<ThemeTokens> }
  );

  return { style, styleVariants };
}
