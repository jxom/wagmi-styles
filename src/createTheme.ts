import { uid } from 'uid';

import type {
  Contract,
  MapLeafNodes,
  ThemeDefinition,
  ThemeVars,
  Tokens,
} from './types';

export function createTheme<ThemeTokens extends Tokens>(
  tokens: ThemeTokens
): [theme: ThemeDefinition<ThemeTokens>, vars: ThemeVars<ThemeTokens>];
export function createTheme<ThemeContract extends Contract>(
  themeContract: ThemeContract,
  tokens: MapLeafNodes<ThemeContract, string | number>
): ThemeDefinition<ThemeContract>;
export function createTheme(arg1: any, arg2?: any) {
  const contract = arg1;
  const tokens = arg2 || arg1;

  const themeId = uid();
  const theme = { _id: themeId, ...tokens };

  if (arg2) {
    return theme;
  }
  return [theme, contract];
}
