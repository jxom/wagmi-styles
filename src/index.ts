export * from './createStyleFn';
export * from './createTheme';
export * from './style';
export * from './ThemeProvider';
export * from './types';
export * from './useStyle';
