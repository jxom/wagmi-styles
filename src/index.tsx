export * from './createStyleFn';
export * from './createTheme';
export * from './ThemeProvider';
export * from './types';
export * from './useStyle';
