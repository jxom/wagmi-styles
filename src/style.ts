import { getStyleSheets } from './private/getStyleSheets';
import type { StylesInput } from './types';

export function style<T>(styles: StylesInput<T>) {
  const { _styleSheet, _composedStyleSheets } = getStyleSheets({
    themeMap: { _: {} },
    getStyles: () => styles,
  });

  return {
    _styleSheet,
    _composedStyleSheets,
  };
}
