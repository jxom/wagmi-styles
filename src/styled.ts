import { useStyle } from './useStyle';
import { style } from './style';

import { generateStyled } from './private/generateStyled';

export const styled = generateStyled({ style, useStyle });
