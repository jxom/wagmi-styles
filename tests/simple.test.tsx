import React from 'react';
import { View } from 'react-native';
import { render } from '@testing-library/react-native';

import { useStyle, style } from '../src';

describe('simple', () => {
  const exampleStyleRef = style({
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    display: 'flex',
  });

  function Example() {
    const exampleStyle = useStyle(exampleStyleRef);
    return <View style={exampleStyle} />;
  }

  it('simple style', () => {
    const { toJSON } = render(<Example />);
    expect(toJSON()).toMatchSnapshot();
  });
});
