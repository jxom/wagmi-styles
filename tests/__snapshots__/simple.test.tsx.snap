// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple simple style 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
      },
    ]
  }
/>
`;
